diff --git a/gst/vaapi/gstvaapipostproc.c b/gst/vaapi/gstvaapipostproc.c
index 052ee4fa..cc143481 100644
--- a/gst/vaapi/gstvaapipostproc.c
+++ b/gst/vaapi/gstvaapipostproc.c
@@ -873,18 +873,23 @@ gst_vaapipostproc_process_vpp (GstBaseTransform * trans, GstBuffer * inbuf,
 
   if (use_vpp_crop (postproc)) {
     crop_rect = &tmp_rect;
-    crop_rect->x = postproc->crop_left;
-    crop_rect->y = postproc->crop_top;
-    crop_rect->width = GST_VIDEO_INFO_WIDTH (&postproc->sinkpad_info)
-        - (postproc->crop_left + postproc->crop_right);
-    crop_rect->height = GST_VIDEO_INFO_HEIGHT (&postproc->sinkpad_info)
-        - (postproc->crop_top + postproc->crop_bottom);
-
-    crop_meta = gst_buffer_get_video_crop_meta (inbuf);
+    crop_meta = gst_buffer_get_video_crop_meta(inbuf);
     if (crop_meta) {
-      crop_rect->x += crop_meta->x;
-      crop_rect->y += crop_meta->y;
+      crop_rect->x = crop_meta->x;
+      crop_rect->y = crop_meta->y;
+      crop_rect->width = crop_meta->width;
+      crop_rect->height = crop_meta->height;
+    } else {
+      crop_rect->x = 0;
+      crop_rect->y = 0;
+      crop_rect->width = GST_VIDEO_INFO_WIDTH(&postproc->sinkpad_info);
+      crop_rect->height = GST_VIDEO_INFO_HEIGHT(&postproc->sinkpad_info);
     }
+
+    crop_rect->x += postproc->crop_left;
+    crop_rect->y += postproc->crop_top;
+    crop_rect->width -= postproc->crop_left + postproc->crop_right;
+    crop_rect->height -= postproc->crop_top + postproc->crop_bottom;
   }
 
   if (!crop_rect)
