diff --git a/gst-libs/gst/vaapi/gstvaapibufferproxy_priv.h b/gst-libs/gst/vaapi/gstvaapibufferproxy_priv.h
index 445201d..a0a7a11 100644
--- a/gst-libs/gst/vaapi/gstvaapibufferproxy_priv.h
+++ b/gst-libs/gst/vaapi/gstvaapibufferproxy_priv.h
@@ -66,8 +66,14 @@ struct _GstVaapiBufferProxy {
   GDestroyNotify        destroy_func;
   gpointer              destroy_data;
   guint                 type;
-  VABufferID            va_buf;
-  VABufferInfo          va_info;
+  union {
+    struct {
+      VABufferID            va_buf;
+      VABufferInfo          va_info;
+    };
+    /* Used for DRM_PRIME_2 */
+    VADRMPRIMESurfaceDescriptor va_descriptor;
+  };
 };
 
 G_GNUC_INTERNAL
diff --git a/gst-libs/gst/vaapi/gstvaapidisplay.h b/gst-libs/gst/vaapi/gstvaapidisplay.h
index 2ca9bd3..54b18dd 100644
--- a/gst-libs/gst/vaapi/gstvaapidisplay.h
+++ b/gst-libs/gst/vaapi/gstvaapidisplay.h
@@ -26,6 +26,7 @@
 #define GST_VAAPI_DISPLAY_H
 
 #include <va/va.h>
+#include <va/va_drmcommon.h>
 #include <gst/gst.h>
 #include <gst/vaapi/gstvaapitypes.h>
 #include <gst/vaapi/gstvaapiprofile.h>
diff --git a/gst/vaapi/gstvaapidecode.c b/gst/vaapi/gstvaapidecode.c
index 9f6eaba..d406b70 100644
--- a/gst/vaapi/gstvaapidecode.c
+++ b/gst/vaapi/gstvaapidecode.c
@@ -270,8 +270,15 @@ gst_vaapidecode_ensure_allowed_srcpad_caps (GstVaapiDecode * decode)
         gst_caps_features_from_string (GST_CAPS_FEATURE_MEMORY_DMABUF));
   }
 #if (GST_VAAPI_USE_GLX || GST_VAAPI_USE_EGL)
-  if (!GST_VAAPI_PLUGIN_BASE_SRC_PAD_CAN_DMABUF (decode)
-      && gst_vaapi_display_has_opengl (GST_VAAPI_PLUGIN_BASE_DISPLAY (decode))) {
+  if (gst_vaapi_mem_type_supports (mem_types,
+          GST_VAAPI_BUFFER_MEMORY_TYPE_DMA_BUF) ||
+      gst_vaapi_mem_type_supports (mem_types,
+          GST_VAAPI_BUFFER_MEMORY_TYPE_DMA_BUF2)) {
+    dma_caps = gst_caps_copy (base_caps);
+    gst_caps_set_features_simple (dma_caps,
+        gst_caps_features_from_string (GST_CAPS_FEATURE_MEMORY_DMABUF));
+  }
+  if (gst_vaapi_display_has_opengl (GST_VAAPI_PLUGIN_BASE_DISPLAY (decode))) {
     gltexup_caps = gst_caps_from_string (GST_VAAPI_MAKE_GLTEXUPLOAD_CAPS);
     if (gltexup_caps) {
       gst_vaapi_caps_set_width_and_height_range (base_caps, min_width,
@@ -546,8 +553,6 @@ is_src_allocator_dmabuf (GstVaapiDecode * decode)
 {
   GstVaapiPluginBase *const plugin = GST_VAAPI_PLUGIN_BASE (decode);
 
-  if (!GST_VAAPI_PLUGIN_BASE_SRC_PAD_CAN_DMABUF (plugin))
-    return FALSE;
   return
       gst_vaapi_is_dmabuf_allocator (GST_VAAPI_PLUGIN_BASE_SRC_PAD_ALLOCATOR
       (plugin));
diff --git a/gst/vaapi/gstvaapipluginbase.c b/gst/vaapi/gstvaapipluginbase.c
index 9b713b5..09201f8 100644
--- a/gst/vaapi/gstvaapipluginbase.c
+++ b/gst/vaapi/gstvaapipluginbase.c
@@ -1017,7 +1017,6 @@ gst_vaapi_plugin_base_decide_allocation (GstVaapiPluginBase * plugin,
       if (gst_structure_get (params, "gst.gl.GstGLContext", GST_TYPE_GL_CONTEXT,
               &gl_context, NULL) && gl_context) {
         gst_vaapi_plugin_base_set_gl_context (plugin, gl_context);
-        gst_vaapi_plugin_base_set_srcpad_can_dmabuf (plugin, gl_context);
         gst_object_unref (gl_context);
       }
     }
@@ -1517,31 +1516,6 @@ gst_vaapi_plugin_base_get_allowed_sinkpad_raw_caps (GstVaapiPluginBase * plugin)
   return plugin->allowed_raw_caps;
 }
 
-/**
- * gst_vaapi_plugin_base_set_srcpad_can_dmabuf:
- * @plugin: a #GstVaapiPluginBase
- * @object: the GL context from gst-gl
- *
- * This function will determine if @object supports dmabuf
- * importing on the base plugin static srcpad.
- *
- * Please note that the context @object should come from downstream.
- **/
-void
-gst_vaapi_plugin_base_set_srcpad_can_dmabuf (GstVaapiPluginBase * plugin,
-    GstObject * object)
-{
-#if GST_VAAPI_USE_EGL && USE_GST_GL_HELPERS
-  GstVaapiPadPrivate *srcpriv = GST_VAAPI_PAD_PRIVATE (plugin->srcpad);
-  GstGLContext *const gl_context = GST_GL_CONTEXT (object);
-
-  srcpriv->can_dmabuf =
-      (!(gst_gl_context_get_gl_api (gl_context) & GST_GL_API_GLES1)
-      && gst_gl_context_check_feature (gl_context,
-          "EGL_EXT_image_dma_buf_import"));
-#endif
-}
-
 static void
 _init_performance_debug (void)
 {
diff --git a/gst/vaapi/gstvaapipluginbase.h b/gst/vaapi/gstvaapipluginbase.h
index 94aa296..9182ed4 100644
--- a/gst/vaapi/gstvaapipluginbase.h
+++ b/gst/vaapi/gstvaapipluginbase.h
@@ -89,8 +89,6 @@ typedef struct _GstVaapiPadPrivate GstVaapiPadPrivate;
   (GST_VAAPI_PLUGIN_BASE_SRC_PAD_PRIVATE(plugin)->caps)
 #define GST_VAAPI_PLUGIN_BASE_SRC_PAD_INFO(plugin) \
   (&GST_VAAPI_PLUGIN_BASE_SRC_PAD_PRIVATE(plugin)->info)
-#define GST_VAAPI_PLUGIN_BASE_SRC_PAD_CAN_DMABUF(plugin) \
-  (GST_VAAPI_PLUGIN_BASE_SRC_PAD_PRIVATE(plugin)->can_dmabuf)
 #define GST_VAAPI_PLUGIN_BASE_SRC_PAD_BUFFER_POOL(plugin) \
   (GST_VAAPI_PLUGIN_BASE_SRC_PAD_PRIVATE(plugin)->buffer_pool)
 #define GST_VAAPI_PLUGIN_BASE_SRC_PAD_ALLOCATOR(plugin) \
@@ -132,8 +130,6 @@ struct _GstVaapiPadPrivate
   guint buffer_size;
   gboolean caps_is_raw;
 
-  gboolean can_dmabuf;
-
   GstAllocator *other_allocator;
   GstAllocationParams other_allocator_params;
 };
@@ -301,11 +297,6 @@ G_GNUC_INTERNAL
 GstCaps *
 gst_vaapi_plugin_base_get_allowed_sinkpad_raw_caps (GstVaapiPluginBase * plugin);
 
-G_GNUC_INTERNAL
-void
-gst_vaapi_plugin_base_set_srcpad_can_dmabuf (GstVaapiPluginBase * plugin,
-    GstObject * object);
-
 G_GNUC_INTERNAL
 gboolean
 gst_vaapi_plugin_copy_va_buffer (GstVaapiPluginBase * plugin,
diff --git a/gst/vaapi/gstvaapipluginutil.c b/gst/vaapi/gstvaapipluginutil.c
index 6fa18f6..d3e37cd 100644
--- a/gst/vaapi/gstvaapipluginutil.c
+++ b/gst/vaapi/gstvaapipluginutil.c
@@ -348,9 +348,7 @@ gst_vaapi_find_gl_context (GstElement * element)
 
   gst_vaapi_find_gl_local_context (element, &plugin->gl_context);
 
-  if (plugin->gl_context) {
-    gst_vaapi_plugin_base_set_srcpad_can_dmabuf (plugin, plugin->gl_context);
-  } else {
+  if (!plugin->gl_context) {
     GstObject *gl_context;
 
     gl_context = gst_vaapi_plugin_base_create_gl_context (plugin);
diff --git a/gst/vaapi/gstvaapipluginutil.h b/gst/vaapi/gstvaapipluginutil.h
index 236ccd3..9eb5234 100644
--- a/gst/vaapi/gstvaapipluginutil.h
+++ b/gst/vaapi/gstvaapipluginutil.h
@@ -117,7 +117,7 @@ gst_vaapi_caps_feature_contains (const GstCaps * caps,
 
 #define GST_VAAPI_MAKE_DMABUF_CAPS                                      \
     GST_VIDEO_CAPS_MAKE_WITH_FEATURES(                                  \
-        GST_CAPS_FEATURE_MEMORY_DMABUF, "{ I420, YV12, RGBA }")
+        GST_CAPS_FEATURE_MEMORY_DMABUF, GST_VAAPI_FORMATS_ALL)
 
 G_GNUC_INTERNAL
 gboolean
diff --git a/gst/vaapi/gstvaapipostproc.c b/gst/vaapi/gstvaapipostproc.c
index 436dbae..0e1b3a8 100644
--- a/gst/vaapi/gstvaapipostproc.c
+++ b/gst/vaapi/gstvaapipostproc.c
@@ -38,6 +38,7 @@
 #include <gst/video/video.h>
 
 #include <gst/vaapi/gstvaapivalue.h>
+#include <gst/vaapi/gstvaapiprofilecaps.h>
 
 #include "gstvaapipostproc.h"
 #include "gstvaapipostprocutil.h"
@@ -1406,9 +1407,7 @@ expand_allowed_srcpad_caps (GstVaapiPostproc * postproc, GstCaps * caps)
   }
   g_value_unset (&value);
 
-  if ((GST_VAAPI_PLUGIN_BASE_SRC_PAD_CAN_DMABUF (postproc)
-          || !gst_vaapi_display_has_opengl (GST_VAAPI_PLUGIN_BASE_DISPLAY
-              (postproc)))
+  if (!gst_vaapi_display_has_opengl (GST_VAAPI_PLUGIN_BASE_DISPLAY (postproc))
       && gl_upload_meta_idx > -1) {
     gst_caps_remove_structure (caps, gl_upload_meta_idx);
   }
-- 
2.25.1

