diff --git a/subprojects/gst-plugins-bad/sys/va/gstvabaseenc.c b/subprojects/gst-plugins-bad/sys/va/gstvabaseenc.c
index c522faf46b70..08ae7f33b6bb 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvabaseenc.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvabaseenc.c
@@ -218,8 +218,9 @@ _get_sinkpad_pool (GstElement * element, gpointer data)
 
   allocator = gst_va_allocator_new (base->display, surface_formats);
 
-  usage_hint = va_get_surface_usage_hint (base->display,
-      VAEntrypointEncSlice, GST_PAD_SINK, FALSE);
+  usage_hint = VA_SURFACE_ATTRIB_USAGE_HINT_GENERIC;
+  /* va_get_surface_usage_hint (display,
+     VAEntrypointEncSlice, GST_PAD_SINK, FALSE); */
 
   base->priv->raw_pool = gst_va_pool_new_with_config (caps, 1, 0, usage_hint,
       GST_VA_FEATURE_AUTO, allocator, &params);
@@ -410,8 +411,9 @@ gst_va_base_enc_propose_allocation (GstVideoEncoder * venc, GstQuery * query)
   if (!caps)
     return FALSE;
 
-  usage_hint = va_get_surface_usage_hint (base->display,
-      VAEntrypointEncSlice, GST_PAD_SINK, gst_video_is_dma_drm_caps (caps));
+  usage_hint = VA_SURFACE_ATTRIB_USAGE_HINT_GENERIC;
+  /* va_get_surface_usage_hint (display,
+     VAEntrypointEncSlice, GST_PAD_SINK, FALSE); */
 
   gst_allocation_params_init (&params);
 
diff --git a/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c b/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c
index 0e11d5b48427..2535fb4d537d 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvaencoder.c
@@ -360,8 +360,9 @@ _create_reconstruct_pool (GstVaDisplay * display, GArray * surface_formats,
 
   gst_video_info_set_format (&info, format, coded_width, coded_height);
 
-  usage_hint = va_get_surface_usage_hint (display,
-      VAEntrypointEncSlice, GST_PAD_SINK, FALSE);
+  usage_hint = VA_SURFACE_ATTRIB_USAGE_HINT_GENERIC;
+  /* va_get_surface_usage_hint (display,
+     VAEntrypointEncSlice, GST_PAD_SINK, FALSE); */
 
   caps = gst_video_info_to_caps (&info);
   gst_caps_set_features_simple (caps,
