diff --git a/subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c b/subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c
index c4ce079d0758..1ba1faef9999 100644
--- a/subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c
+++ b/subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c
@@ -845,15 +845,9 @@ gst_msdkdec_set_src_caps (GstMsdkDec * thiz, gboolean need_allocation)
 
   if (need_allocation) {
     /* Find allocation width and height */
-    width =
-        GST_ROUND_UP_16 (thiz->param.mfx.FrameInfo.Width ? thiz->param.mfx.
-        FrameInfo.Width : GST_VIDEO_INFO_WIDTH (&thiz->output_state->info));
-    height =
-        GST_ROUND_UP_32 (thiz->param.mfx.FrameInfo.Height ? thiz->param.mfx.
-        FrameInfo.Height : GST_VIDEO_INFO_HEIGHT (&thiz->output_state->info));
-
-    /* set allocation width and height in allocation_caps,
-     * which may or may not be similar to the output_state caps */
+    width = GST_VIDEO_INFO_WIDTH (&thiz->output_state->info);
+    height = GST_VIDEO_INFO_HEIGHT (&thiz->output_state->info);
+
     allocation_caps = gst_caps_copy (thiz->output_state->caps);
     gst_caps_set_simple (allocation_caps, "width", G_TYPE_INT, width, "height",
         G_TYPE_INT, height, NULL);
