diff --git a/subprojects/gst-plugins-bad/sys/va/gstvafilter.c b/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
index c35ba040cf23..1632906a736c 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
@@ -1827,12 +1827,6 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
 
   dpy = gst_va_display_get_va_dpy (self->display);
 
-  status = vaBeginPicture (dpy, self->context, out_surface);
-  if (status != VA_STATUS_SUCCESS) {
-    GST_ERROR_OBJECT (self, "vaBeginPicture: %s", vaErrorStr (status));
-    return FALSE;
-  }
-
   sample = tx->next (tx->user_data);
   for (; sample; sample = tx->next (tx->user_data)) {
     VAProcPipelineParameterBuffer params = { 0, };
@@ -1853,7 +1847,6 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       .surface = in_surface,
       .surface_region = &sample->input_region,
       .output_region = &sample->output_region,
-      .output_background_color = 0xff000000,
       .filter_flags = self->scale_method | self->interpolation_method,
     };
     /* *INDENT-ON* */
@@ -1871,6 +1864,12 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       params.blend_state = &blend;
     }
 
+    status = vaBeginPicture (dpy, self->context, out_surface);
+    if (status != VA_STATUS_SUCCESS) {
+      GST_ERROR_OBJECT (self, "vaBeginPicture: %s", vaErrorStr (status));
+      return FALSE;
+    }
+
     status = vaCreateBuffer (dpy, self->context,
         VAProcPipelineParameterBufferType, sizeof (params), 1, &params,
         &buffer);
@@ -1885,12 +1884,12 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       GST_ERROR_OBJECT (self, "vaRenderPicture: %s", vaErrorStr (status));
       goto fail_end_pic;
     }
-  }
 
-  status = vaEndPicture (dpy, self->context);
-  if (status != VA_STATUS_SUCCESS) {
-    GST_ERROR_OBJECT (self, "vaEndPicture: %s", vaErrorStr (status));
-    return FALSE;
+    status = vaEndPicture (dpy, self->context);
+    if (status != VA_STATUS_SUCCESS) {
+      GST_ERROR_OBJECT (self, "vaEndPicture: %s", vaErrorStr (status));
+      return FALSE;
+    }
   }
 
   return TRUE;
