diff --git a/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c b/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c
index ee709d4de..5e0afe660 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c
@@ -1665,7 +1665,7 @@ gst_va_compositor_class_init (gpointer g_class, gpointer class_data)
    */
   if (GST_VA_DISPLAY_IS_IMPLEMENTATION (display, INTEL_IHD)) {
     g_object_class_install_property (object_class,
-        GST_VA_FILTER_PROP_SCALE_METHOD,
+        PROP_SCALE_METHOD,
         g_param_spec_enum ("scale-method", "Scale Method",
             "Scale method to use", GST_TYPE_VA_SCALE_METHOD,
             VA_FILTER_SCALING_DEFAULT,
