diff --git a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvautils.c b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvautils.c
index 101267e0b..f9dfdcb2f 100644
--- a/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvautils.c
+++ b/subprojects/gst-plugins-bad/gst-libs/gst/va/gstvautils.c
@@ -315,14 +315,21 @@ gst_va_handle_set_context (GstElement * element, GstContext * context,
   }
 
   if (display_replacement) {
+
+     const gchar *tlv_vadpy_str = getenv("threshold_limit_value_vadpy");
+     gchar *end_ptr;
+
+     gint tlv_vadpy_int = (gint)g_ascii_strtoll(tlv_vadpy_str? tlv_vadpy_str: "", &end_ptr, 10);
+     tlv_vadpy_int = (!tlv_vadpy_str || *end_ptr != '\0') ? 0 : (tlv_vadpy_int < 0 ? 0 : (tlv_vadpy_int > 8 ? 8 : tlv_vadpy_int));
+
 #ifdef HAVE_LIBDRM
-    if (!gst_va_display_drm_check_i915(display_replacement)) {
+    if (!gst_va_display_drm_check_i915(display_replacement) && tlv_vadpy_int) {
       if (gst_va_display_is_implementation (display_replacement,
           GST_VA_IMPLEMENTATION_INTEL_IHD) && !from_neighbor) {
           guint ref;
 
           ref = g_atomic_int_get (&((GObject *) display_replacement)->ref_count);
-          if (ref > 4) {
+          if (ref > tlv_vadpy_int) {
             gst_object_unref (display_replacement);
           return FALSE;
           }
