diff --git a/subprojects/gst-plugins-bad/sys/va/gstvaav1enc.c b/subprojects/gst-plugins-bad/sys/va/gstvaav1enc.c
index 51ab98776..b822edc5e 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvaav1enc.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvaav1enc.c
@@ -3588,6 +3588,7 @@ _av1_fill_frame_header (GstVaAV1Enc * self,
     .allow_screen_content_tools = 0,
     .frame_size_override_flag = 0,
     .frame_width = self->sequence_hdr.max_frame_width_minus_1 + 1,
+    .upscaled_width = self->sequence_hdr.max_frame_width_minus_1 + 1,
     .frame_height = self->sequence_hdr.max_frame_height_minus_1 + 1,
     .order_hint = pic_param->order_hint,
     .primary_ref_frame = pic_param->primary_ref_frame,
