diff --git a/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c b/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c
index 0bd4805f55..21c0b3d32a 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvacompositor.c
@@ -1664,7 +1664,7 @@ gst_va_compositor_class_init (gpointer g_class, gpointer class_data)
    */
   if (GST_VA_DISPLAY_IS_IMPLEMENTATION (display, INTEL_IHD)) {
     g_object_class_install_property (object_class,
-        GST_VA_FILTER_PROP_SCALE_METHOD,
+        PROP_SCALE_METHOD,
         g_param_spec_enum ("scale-method", "Scale Method",
             "Scale method to use", GST_TYPE_VA_SCALE_METHOD,
             VA_FILTER_SCALING_DEFAULT,
diff --git a/subprojects/gst-plugins-bad/sys/va/gstvafilter.c b/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
index c35ba040cf..06102247c0 100644
--- a/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
+++ b/subprojects/gst-plugins-bad/sys/va/gstvafilter.c
@@ -62,6 +62,7 @@ struct _GstVaFilter
 
   guint32 scale_method;
   guint32 interpolation_method;
+  guint32 pipeline_flags;
 
   gboolean crop_enabled;
 
@@ -970,7 +971,13 @@ gst_va_filter_set_scale_method (GstVaFilter * self, guint32 method)
   g_return_val_if_fail (GST_IS_VA_FILTER (self), FALSE);
 
   GST_OBJECT_LOCK (self);
+  guint32 pipeline_flags = 0;
+  if (method == VA_FILTER_SCALING_FAST) {
+    pipeline_flags = VA_PROC_PIPELINE_FAST;
+  }
+
   self->scale_method = method;
+  self->pipeline_flags = pipeline_flags;
   GST_OBJECT_UNLOCK (self);
 
   return TRUE;
@@ -1672,6 +1679,7 @@ _create_pipeline_buffer (GstVaFilter * self, GstVaSample * src,
     .input_color_properties = self->input_color_properties,
     .output_color_properties = self->output_color_properties,
     .filter_flags = self->scale_method | self->interpolation_method,
+    .pipeline_flags = self->pipeline_flags,
     /* output to SDR */
     .output_hdr_metadata = NULL,
   };
@@ -1827,12 +1835,6 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
 
   dpy = gst_va_display_get_va_dpy (self->display);
 
-  status = vaBeginPicture (dpy, self->context, out_surface);
-  if (status != VA_STATUS_SUCCESS) {
-    GST_ERROR_OBJECT (self, "vaBeginPicture: %s", vaErrorStr (status));
-    return FALSE;
-  }
-
   sample = tx->next (tx->user_data);
   for (; sample; sample = tx->next (tx->user_data)) {
     VAProcPipelineParameterBuffer params = { 0, };
@@ -1853,8 +1855,8 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       .surface = in_surface,
       .surface_region = &sample->input_region,
       .output_region = &sample->output_region,
-      .output_background_color = 0xff000000,
       .filter_flags = self->scale_method | self->interpolation_method,
+      .pipeline_flags = self->pipeline_flags,
     };
     /* *INDENT-ON* */
     GST_OBJECT_UNLOCK (self);
@@ -1871,6 +1873,12 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       params.blend_state = &blend;
     }
 
+    status = vaBeginPicture (dpy, self->context, out_surface);
+    if (status != VA_STATUS_SUCCESS) {
+      GST_ERROR_OBJECT (self, "vaBeginPicture: %s", vaErrorStr (status));
+      return FALSE;
+    }
+
     status = vaCreateBuffer (dpy, self->context,
         VAProcPipelineParameterBufferType, sizeof (params), 1, &params,
         &buffer);
@@ -1885,12 +1893,12 @@ gst_va_filter_compose (GstVaFilter * self, GstVaComposeTransaction * tx)
       GST_ERROR_OBJECT (self, "vaRenderPicture: %s", vaErrorStr (status));
       goto fail_end_pic;
     }
-  }
 
-  status = vaEndPicture (dpy, self->context);
-  if (status != VA_STATUS_SUCCESS) {
-    GST_ERROR_OBJECT (self, "vaEndPicture: %s", vaErrorStr (status));
-    return FALSE;
+    status = vaEndPicture (dpy, self->context);
+    if (status != VA_STATUS_SUCCESS) {
+      GST_ERROR_OBJECT (self, "vaEndPicture: %s", vaErrorStr (status));
+      return FALSE;
+    }
   }
 
   return TRUE;
